Adding GetByID function

morses-code opened the pull request:
### What
Added `GetByID()` function

### How to review
Check tests pass and review code for any obvious errors.

morses-code committed:
Changes to client

- Adding interfaces for redis

morses-code committed:
Needed to add Ping to interface

morses-code committed:
Some name changing. Resolving conflicts and struct with same name as pkg

morses-code committed:
Updated return types on interface

morses-code committed:
Added tests for setting up new redis client

morses-code committed:
Lots of refactoring, added mocks and tests

- Updated package name from redis to dpredis
- Removed `Resulter` interface
- Moved and renamed Clienter interface
- Added a return when the `client.Set()` func fails
- Added tests to cover valid and invalid sessions and for when the `Set()` func for redis errors
- Added interfaces package

morses-code committed:
Adding GetByID() function

- Created GetByID() function 
- Added `Get()` function to the RedisClienter interface
- Added `Get()` function implementation from go-redis
- Added tests for `GetByID()`
- Updated mock to include `Get()` function

daiLlew commented on client.go:
```diff
@@ -1,20 +1,27 @@
 var ctx = context.Background()
 
-// Client - structure for the redis client
+// RedisClient - structure for the redis client
```
☝️ the `ctx` should be passed into each public method of the client by the calling code.

morses-code replied:
Done

daiLlew commented on client.go:
```diff
@@ -37,33 +44,72 @@ func NewClient(c Config) (*Client, error) {
 }
 
-//Set add session to redis
+// Set - add session to redis
```
👇 We have an unofficial coding standard whereby the first argument for a client should always be `context`.

daiLlew commented on client.go:
```diff
@@ -37,33 +44,72 @@ func NewClient(c Config) (*Client, error) {
+func (c *Client) GetByID(id string) (*session.Session, error) {
+	if id == "" {
+		log.Event(ctx, "id value is blank", log.ERROR)
+		return nil, errors.New("id value is blank")
```
Consider making this error a public const - it allows the caller to "switch" on the error type if they need to. Also consider making the message more descriptive for example `session id required by was empty` gives you a much clearer idea of the error without having to dig into the code.

morses-code replied:
Made a few of the error messages constants.

daiLlew commented on client.go:
```diff
@@ -37,33 +44,72 @@ func NewClient(c Config) (*Client, error) {
+
+	err = json.Unmarshal([]byte(msg), &s)
+	if err != nil {
+		log.Event(ctx, "failed to unmarshal session", log.Error(err), log.ERROR)
```
`failed to unmarshal get session json response` - more descriptive.

daiLlew commented on interfaces/interfaces.go:
```diff
@@ -0,0 +1,16 @@
+
+// RedisClienter - interface for redis
+type RedisClienter interface {
+	Set(string, interface{}, time.Duration) *redis.StatusCmd
+	Get(string) *redis.StringCmd
```
Add parameter names.

morses-code replied:
Names have now been added.

daiLlew commented on client_test.go:
```diff
@@ -1,47 +1,184 @@
+	})
+}
+
+func setUpClient(addr, password string, database int, ttl time.Duration) (*Client, error) {
+	c, err := NewClient(Config{
+		Addr:     addr,
+		Password: password,
+		Database: database,
+		TTL:      ttl,
+	})
+	return c, err
```
This method isn't adding any functionality - remove it and use the `NewClient` function directly.

morses-code replied:
Agreed, no idea why I did this! Removed.

daiLlew commented on client_test.go:
```diff
@@ -1,47 +1,184 @@
+
+func TestClient_GetByID(t *testing.T) {
+	Convey("Given a session ID", t, func() {
+		mockRedisClient, mockClient := setUpMocks(*redis.NewStatusCmd(), *redis.NewStringResult(string(resp), nil))
```
I'm being picky name but `mockClient` is not a mock - its the target of this test. Rename it to `client` so its clear this is the struct being tested.

morses-code replied:
Name has been changed.

daiLlew commented on client_test.go:
```diff
@@ -1,47 +1,184 @@
+			s, err := mockClient.GetByID("1234")
+			So(err, ShouldBeNil)
+
+			msg, err := mockRedisClient.Get(s.ID).Result()
+			So(err, ShouldBeNil)
```
```
Convey("Then redis client Get is called 1 time with the expected args", func(){
    So(mockRedisClient. GetCalls(), shouldHaveLength, 1)
})
``` 
etc

daiLlew commented on client_test.go:
```diff
@@ -1,47 +1,184 @@
+			So(err, ShouldBeNil)
+
+			Convey("Then session is returned", func() {
+				So(storedSession.ID, ShouldEqual, s.ID)
```
formatting?

morses-code replied:
Turns out a lot was weird with how I wrote the `GetByID()` tests, gone through them and removed code that was duplicating the mocks behaviour.

morses-code committed:
Attempt to create interface for redis.StatusCMD

morses-code committed:
Revert "Attempt to create interface for redis.StatusCMD"

This reverts commit 8012f7ae16a8c7a7074f4840d1e5090ea1c70207.

morses-code committed:
Auto stash before revert of "Attempt to create interface for redis.StatusCMD"

morses-code committed:
Ping() reverted to *StatusCmd return type.

morses-code committed:
Added parameter names on interface, updated mock

morses-code committed:
Added validation for TTL and new test

morses-code committed:
Merge branch 'feature/add-set-func' into feature/add-getbyid-func

morses-code committed:
Addressing PR comments

- Added err constants
- General formatting improvements in client.go
- Client functions now require context to be passed in
- Removed unnecessary function in client_test
- Added context for tests
- Improved tests for GetByID()
- Updated interface so parameters are descriptive

daiLlew commented on client.go:
```diff
@@ -36,34 +48,77 @@ func NewClient(c Config) (*Client, error) {
+// GetByID - gets a session from redis using its ID
+func (c *Client) GetByID(ctx context.Context, id string) (*session.Session, error) {
+	if id == "" {
+		log.Event(ctx, ErrEmptySessionID, log.ERROR)
```
generally speaking you should either log or handle the error. In this case I would remove the log and just return it. Otherwise you end up with multiple points repeating the same error.

daiLlew commented on client.go:
```diff
@@ -1,20 +1,32 @@
 
-var ctx = context.Background()
+const (
+	ErrEmptySessionID    = "session id required but was empty"
+	ErrEmptySession      = "session is empty"
+	ErrFailedToUnmarshal = "failed to unmarshal get session json response"
+	ErrFailedToMarshal   = "failed to marshal session"
```
Make these type `error`.

daiLlew commented on client.go:
```diff
@@ -36,34 +48,77 @@ func NewClient(c Config) (*Client, error) {
+	sJSON, err := s.MarshalJSON()
 	if err != nil {
-		log.Event(ctx, "failed to marshal session", log.Error(err), log.ERROR)
+		log.Event(ctx, ErrFailedToMarshal, log.Error(err), log.ERROR)
```
ditto

daiLlew commented on client.go:
```diff
@@ -36,34 +48,77 @@ func NewClient(c Config) (*Client, error) {
+	msg, err := c.client.Set(s.ID, sJSON, c.ttl).Result()
 	if err != nil {
 		log.Event(ctx, msg, log.Error(err), log.ERROR)
+		return err
```
ditto

morses-code committed:
Updated errors and removed logging and context

morses-code committed:
Updated errors, removed context and updated tests

morses-code committed:
Merge branch 'feature/add-set-func' into feature/add-getbyid-func

morses-code committed:
Updated wording of tests

- Tests should now be more clear

morses-code committed:
Errors more descriptive, wrapped error returned from redis client

morses-code committed:
Fix tests with new errors

morses-code committed:
Merge branch 'feature/add-set-func' into feature/add-getbyid-func

eldeal commented on interfaces/interfaces.go:
```diff
@@ -0,0 +1,16 @@
+package interfaces
```
Is there a reason this interface needs to be in its own package? I would expect interfaces to live as close as possible to where they're used - so the top level of this repo alongside client.go and client_test.go

It would be reasonable/expected for this to be in the client.go file alongside the Client type definitions, though keeping it in its own file isnt a problem. 

It's reasonable to put the mocks in their own directory/package though 👍

morses-code replied:
The interfaces are now in the same package as client, had to do the same with the mocks. Had a cirular import dependancy that I couldn't figure out without that in the same package too.

eldeal commented on client.go:
```diff
@@ -1,20 +1,32 @@
+// Client - structure for the cache client
 type Client struct {
-	client *goredis.Client
+	client RedisClient
```
Is having a type for this really necessary? i'm not sure what value it is adding, when the interface itself could be used for this type. You could implement the necessary functions off the Client type, without the extra hop through RedisClient

morses-code replied:
Yup, agree. Looking at it again it's a bit convoluted having this too. I think initially it had some other stuff in there too but it had been reworked.

eldeal commented on client.go:
```diff
@@ -1,20 +1,32 @@
+	"fmt"
 	"time"
 
+	. "github.com/ONSdigital/dp-redis/interfaces"
```
Try and avoid anonymous imports :) The only place we tend to use them is in tests when importing go-convey, and as a rule I'd say we probably only ever want to use them in tests. For production code, we always want to know exactly what is in this package and what we're having to pull from somewhere else

morses-code replied:
Gone with the above change. 👍

morses-code committed:
Pulled interfaces.go into the main package, same with the mock. 

- Updated the tests to reflect the refactor

morses-code committed:
Merge branch 'develop' into feature/add-getbyid-func

morses-code committed:
go mod updates
